src/codegen/function.rs

```diff
@@ -187,3 +294,88 @@ pub fn body_chunk(env: &Env, analysis: &analysis::functions::Info) -> Chunk {
 
     builder.generate(env)
 }
+
+pub fn body_chunk_futures(env: &Env, analysis: &analysis::functions::Info) -> StdResult<String, fmt::Error> {
+    use std::fmt::Write;
+    use analysis::bounds::BoundType;
+    use analysis::ref_mode::RefMode;
+
+    let mut body = String::new();
+
+    if env.config.library_name != "Gio" {
+        try!(writeln!(body, "use gio::GioFuture;"));
+    } else {
+        try!(writeln!(body, "use GioFuture;"));
+    }
+    try!(writeln!(body, "use send_cell::SendCell;"));
+    try!(writeln!(body, ""));
+
+    for par in analysis.parameters.rust_parameters.iter().skip(1) {
+        if par.name == "cancellable" || par.name == "callback" {
+            continue;
+        }
+
+        let c_par = &analysis.parameters.c_parameters[par.ind_c];
+
+        let bounds = analysis.bounds.get_parameter_alias_info(&par.name);
+        let is_into = if let Some((_, BoundType::Into(..))) = bounds { true } else { false };
+
+        let type_ = env.type_(par.typ);
+        let is_str = if let library::Type::Fundamental(library::Fundamental::Utf8) = *type_ { true } else { false };
+
+        if is_into {
+            try!(writeln!(body, "let {} = {}.into();", par.name, par.name));
+            if is_str {
+                try!(writeln!(body, "let {} = {}.map(String::from);", par.name, par.name));
+            } else if c_par.nullable.0 {
+                try!(writeln!(body, "let {} = {}.cloned();", par.name, par.name));
+            }
+        } else if is_str {
+            try!(writeln!(body, "let {} = String::from({});", par.name, par.name));
+        } else if c_par.ref_mode != RefMode::None {
+            try!(writeln!(body, "let {} = {}.clone();", par.name, par.name));
+        }
+    }
+
+    try!(writeln!(body, "GioFuture::new(self, move |obj, send| {{"));
```

Comment:
This would currently not work if the async function is a global one (https://github.com/gtk-rs/gir/pull/581) and also not for methods not taking `self` as first argument (`gdk_pixbuf::new_from_stream_async`).

Any suggestions would be welcome how to make this more generic. I guess the object could become an `Option` for the `GioFuture`, only stored internally (we need to keep it alive until the operation is done), and the object is then passed via a `SendCell` directly into the closure.